File: client/src/controller.js
export default class Controller {
  constructor(eventClient, view) {
    this.eventClient = eventClient;
    this.view = view;
  }

  start() {
    var self = this;
    this.eventClient.on('position', function(e, p){
      self.onPositionEvent(e, p);
    });
    return this;
  }

  onInited(position, speed, screenSize) {
    this.eventClient.send("position", Point.toPayload(position));
    this.eventClient.send("speed", Point.toPayload(speed));
    this.eventClient.send("screenSize", Point.toPayload(screenSize));
  }

  loop() {

  }

  onPositionEvent(_, payload) {
    let p = Point.fromPayload(payload);
    let cat = this.view.cats[0];
    if (cat) {
      cat.x = p.x;
      cat.y = p.y;
    }
  }
}

class Point {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }

  static fromPayload(payload) {
    let arr = payload.split(',');
    let x = parseInt(arr[0]);
    let y = parseInt(arr[1]);
    return new Point(x, y);
  }

  static toPayload(pointLike) {
    return `${pointLike.x},${pointLike.y}`;
  }
}

File: client/src/main/scala/Controller.scala
import scala.scalajs.js

class Controller(eventClient: EventSocketClient, view: View) {
  var index: String = _
  val tmpPoint = new Point(0, 0)

  def start() = {
    eventClient.on("created", onCreated _)
    eventClient.on("gameSnapshot", onGameSnapshot _)
    eventClient.on("enemyConnected", onEnemyConnected _)
    eventClient.on("enemyDisconnected", onEnemyDisconnected _)
  }

  def onInited(screenSize: Point): Unit = {

  }

  def loop() = {

  }

  def onCreated(evType: String, payload: String) = {
    index = addCreature(payload)
  }

  def onGameSnapshot(evType: String, payload: String) = {
    payload.split(";").foreach {
      part =>
        val arr = part.split(":")
        val index = arr(0)
        var position = Point.fromPayload(arr(1), tmpPoint)
        view.catsByIndex.get(index) match {
          case Some(cat) =>
            cat.position.copyFrom(position)
          case None =>
        }
    }
  }

  def onEnemyConnected(evType: String, payload: String) = {
    addCreature(payload)
  }

  def onEnemyDisconnected(evType: String, payload: String) = {
    val (index, _, _, _) = deseriaizedData(payload)
    view.cats.find(_.index == index) match {
      case Some(disconnected) =>
        view.removeCat(disconnected)
      case None =>
    }
  }

  private def addCreature(payload: String): String = {
    val (index, position, speed, size) = deseriaizedData(payload)
    view.addCat(new Cat(view = null, index = index, position = position, speed = speed, size = size))
    index
  }

  private def deseriaizedData(payload: String): (String, Point, Point, Point) = {
    val parts = payload.split(":")
    (parts(0),
      Point.fromPayload(parts(1)),
      Point.fromPayload(parts(2)),
      Point.fromPayload(parts(3)))
  }
}
